crates/network-scanner-proto/src/icmp_v4.rs

```diff
@@ -0,0 +1,387 @@
+                bytes.append(&mut buf);
+                bytes.extend_from_slice(payload);
+            }
+            Self::ParameterProblem {
+                // type 12
+                pointer,
+                padding,
+                header,
+            } => {
+                bytes.push(*pointer);
+                bytes.push(padding.0);
+                let mut buf = vec![0, 2];
+                BigEndian::write_u16(&mut buf, padding.1);
+                bytes.append(&mut buf);
+                bytes.extend_from_slice(header);
+            }
+            Self::Timestamp {
+                // type 13
+                identifier,
+                sequence,
+                originate,
+                receive,
+                transmit,
+            }
+            | Self::TimestampReply {
+                // type 14
+                identifier,
+                sequence,
+                originate,
+                receive,
+                transmit,
+            } => {
+                let mut buf = vec![0, 2];
+                BigEndian::write_u16(&mut buf, *identifier);
+                bytes.append(&mut buf);
+                BigEndian::write_u16(&mut buf, *sequence);
+                bytes.append(&mut buf);
+                buf = vec![0, 4];
+                BigEndian::write_u32(&mut buf, *originate);
+                bytes.append(&mut buf);
+                BigEndian::write_u32(&mut buf, *receive);
+                bytes.append(&mut buf);
+                BigEndian::write_u32(&mut buf, *transmit);
+                bytes.append(&mut buf);
+            }
+            Self::Information {
+                // type 15
+                identifier,
+                sequence,
+            }
+            | Self::InformationReply {
+                // type 16
+                identifier,
+                sequence,
+            } => {
+                let mut buf = vec![0, 2];
+                BigEndian::write_u16(&mut buf, *identifier);
+                bytes.append(&mut buf);
+                BigEndian::write_u16(&mut buf, *sequence);
+                bytes.append(&mut buf);
+            }
+        }
+        bytes
+    }
+}
+pub struct Icmpv4Packet {
+    pub code: u8,
+    pub checksum: u16,
+    pub message: Icmpv4Message,
+}
+
+impl TryFrom<&[u8]> for Icmpv4Packet {
+    type Error = PacketParseError;
+
+    fn try_from(value: &[u8]) -> Result<Self, Self::Error> {
+        Self::parse(value)
+    }
+}
+
+impl Into<Vec<u8>> for Icmpv4Packet {
+    fn into(self) -> Vec<u8> {
+        self.to_bytes(true)
+    }
+}
+
+impl Icmpv4Packet {
+    pub fn parse(bytes: impl AsRef<[u8]>) -> Result<Self, PacketParseError> {
+        let mut bytes = bytes.as_ref();
+        let mut packet_len = bytes.len();
+        if bytes.len() < 28 {
+            return Err(PacketParseError::PacketTooSmall(packet_len));
+        }
+        // NOTE(jwall) Because we use raw sockets the first 20 bytes are the IPv4 header.
```

Comment:
question: Who / what is jwall? Did you copy paste code from somewhere else?

Reply:
yes, I copy and pasted code from a MIT licensed library, he did not have everything I wantted so I copy and modified it a little bit.

Reply:
https://github.com/zaphar/icmp-socket

Reply:
Okay. The repository has not been updated for two years now, so it may not be maintained anymore. Otherwise sending a PR to add the missing bits is the way to go.

However, even if the Apache 2 is indeed permissive, you can’t just take the code without the original license like this.
I encourage you to peruse the licensing conditions: https://www.apache.org/licenses/LICENSE-2.0

First, you can’t change the license of the existing code, because Devolutions is not the copyright holder.
In the past, we [directly asked the copyright holder of a project we forked in order to re-license it](https://github.com/KizzyCode/serde_asn1_der-rust/issues/1#issuecomment-559145397).
For this reason, you need to copy the the [LICENSE](https://github.com/zaphar/icmp-socket/blob/master/LICENSE) and put it in `network-scanner-proto`. Also add the [`license` key in the Cargo.toml](https://github.com/zaphar/icmp-socket/blob/02b0714302db717936be987f006edba32cdf0d38/Cargo.toml#L8).
Usually, we would add a line to add ourselves to the file, but I see the original author didn’t modify it to add themselves, so I’m unsure if we should in this case.

Also [add the original author in the `authors` key](https://github.com/zaphar/icmp-socket/blob/02b0714302db717936be987f006edba32cdf0d38/Cargo.toml#L6).

In addition to that, add a README with a notice which makes it clear we include code from the other project:
```
This project includes portions of [icmp-socket](https://github.com/zaphar/icmp-socket), held by Jeremy Wall, 2022.
```

Next time, don’t forget to mention when you copy code from another project. It’s critical to respect the licensing conditions.

Reply:
Thank you very much. Will do better next time.